apps/kg/models/mxnet/score_fun.py

```diff
@@ -46,6 +57,125 @@ def fn(heads, relations, tails, num_chunks, chunk_size, neg_sample_size):
+                # pos node, project to its relation
+                projection = self.projection_emb(rel_id, gpu_id, trace)
+                projection = projection.reshape(-1, self.entity_dim, self.relation_dim)
+                tail = tail.reshape(-1, 1, self.entity_dim)
+                tail = nd.batch_dot(tail, projection)
+                tail = tail.reshape(num_chunks, -1, self.relation_dim)
+
+                # neg node, each project to all relations
+                projection = projection.reshape(num_chunks, -1, self.entity_dim, self.relation_dim)
+                head = head.reshape(num_chunks, -1, 1, self.entity_dim)
+                num_rels = projection.shape[1]
+                num_nnodes = head.shape[1]
+
+                heads = []
+                for i in range(num_chunks):
+                    head_negs = []
+                    for j in range(num_nnodes):
+                        head_neg = head[i][j]
+                        head_neg = head_neg.reshape(1, 1, self.entity_dim)
+                        head_neg = nd.broadcast_axis(head_neg, axis=0, size=num_rels)
+                        head_neg = nd.batch_dot(head_neg, projection[i])
+                        head_neg = head_neg.squeeze(axis=1)
+                        head_negs.append(head_neg)
+                    head_negs = nd.stack(*head_negs, axis=1)
+                    heads.append(head_negs)
+                head = nd.stack(*heads)
+                return head, tail
+            return fn
+        else:
+            def fn(rel_id, num_chunks, head, tail, gpu_id, trace=False):
+                # pos node, project to its relation
+                projection = self.projection_emb(rel_id, gpu_id, trace)
+                projection = projection.reshape(-1, self.entity_dim, self.relation_dim)
+                head = head.reshape(-1, 1, self.entity_dim)
+                head = nd.batch_dot(head, projection).squeeze()
+                head = head.reshape(num_chunks, -1, self.relation_dim)
+
+                projection = projection.reshape(num_chunks, -1, self.entity_dim, self.relation_dim)
+                tail = tail.reshape(num_chunks, -1, 1, self.entity_dim)
+                num_rels = projection.shape[1]
+                num_nnodes = tail.shape[1]
+
+                tails = []
+                for i in range(num_chunks):
+                    tail_negs = []
+                    for j in range(num_nnodes):
+                        tail_neg = tail[i][j]
+                        tail_neg = tail_neg.reshape(1, 1, self.entity_dim)
+                        tail_neg = nd.broadcast_axis(tail_neg, axis=0, size=num_rels)
+                        tail_neg = nd.batch_dot(tail_neg, projection[i])
+                        tail_neg = tail_neg.squeeze(axis=1)
+                        tail_negs.append(tail_neg)
+                    tail_negs = nd.stack(*tail_negs, axis=1)
+                    tails.append(tail_negs)
+                tail = nd.stack(*tails)
```

Comment:
you are iterating over individual nodes. how about projecting negative nodes to edges? in this way, all computation can be batched. It might have better performance.

Reply:
Iterating because I want to save memory with MXNet backend. MXNet do not provide dot operation supporting broadcast. 
If we want to project nodes to edges, we should build the neg-head and neg-tail graph from pos and neg graph. This should break most of our code I think. BTW, in MXNet 1.6, deep numpy will provide einsum API, then we can simplify this code.